tf_encrypted/protocol/pond/pond.py

```diff
@@ -295,7 +298,72 @@ def define_public_variable(
         x_on_1 = factory.variable(v_on_1)
 
     x = PondPublicVariable(self, x_on_0, x_on_1, apply_scaling)
-    _initializers.append(x.initializer)
+    return x
+
+  def define_private_tensor(
```

Comment:
I've been pondering this guy for a while. What was the motivation to introduce it here?

Reply:
Since `tf.function`s can be called multiple times while they're tracing and it only wants `tf.Variable`s to be initialized only once you need to have a check in your `tf.function` to make sure they're not initialized more than once. The recommended way is to the store the Variables inside of a class that created outside of the `tf.function`. Here's the example for the [docs](https://www.tensorflow.org/tutorials/customization/performance#variables):

```
class C: pass
obj = C(); obj.v = None

@tf.function
def g(x):
  if obj.v is None:
    obj.v = tf.Variable(1.0)
  return obj.v.assign_add(x)

print(g(1.0))  # 2.0
print(g(2.0))  # 4.0
```

So I just decided to make a `define_private_tensor` for now until we have a better solution. But also its not clear to my why we're using Variables all over the place when they're not really needed for a lot of our tests.

Reply:
> So I just decided to make a `define_private_tensor` for now until we have a better solution. But also its not clear to my why we're using Variables all over the place when they're not really needed for a lot of our tests.

It makes sense to have it for now, but I think there's a larger discussion to be had before making this part of the public API.

The deeper issue is that TFE has mostly been designed around the idea that the script itself is simply setting up an encrypted computation between a set of parties, resulting in a graph that is safe for distribution by only describing how the computation is to be performed without containing any sensitive information. This fits with what we explicitly called the `master` earlier. The way of expressing the need to get data into the computation is via `tfe.local_computation` and `tfe.define_private_placeholder`.

However, in tests and [even some examples](https://github.com/tf-encrypted/tf-encrypted/tree/master/examples/notebooks/serving) it has proven useful or intuitive to sometimes treat the script as an active party in the computation, for instance as the client in an encrypted computation instead. In these cases the generated graph may contain sensitive values and is in general *not* safe for distribution. The current API tries to prevent this, resulting it the work around via variables that feels like a misuse.

On a technically level, the issue is that TFE is currently designed around in-graph computations with limited support for between-graph computations.

Moving forward we will need to find a way to accommodate both use cases, but I've previously avoided introducing a function like `define_private_tensor` since it may be confusing to the user that the value is *not* guaranteed to remain private. Two approaches for now might be:

1. update the examples (and perhaps even the tests) to actually do the right thing and provide inputs via `tfe.local_computation`

2. or, keep `define_private_tensor` but make a big fat note that this needs to be further addressed

Reply:
Agree with all that. Did not realize were using variables for a work around like this so feel like something needs to change with that API, guess that's what you're proposing with `tfe.local_computation`? Could we just get rid of `define_private_tensor` and `define_private_variable` (as public APIs anyway)?

Reply:
> feel like something needs to change with that API, guess that's what you're proposing with `tfe.local_computation`?

We need to figure out how to be better at between-graph computations, as that'll likely be needed for TF Federated and PySyft-TF as well. I perhaps see this in addition in the existing `tfe.local_computation` though.

> Could we just get rid of `define_private_tensor` and `define_private_variable` (as public APIs anyway)?

We probably need to keep some notion of private variable around, although I think the API should be changed to something like `tfe.Variable`. `define_private_tensor` could go if there's an alternative way for you to quickly fix up the examples.

Reply:
> like `tfe.Variable.define_private_tensor` could go if there's an alternative way for you to quickly fix up the examples.

Lost me here. Typo?

Reply:
I mean that we should probably keep `tfe.define_private_variable` but rename it to `tfe.Variable` to match better with TensorFlow (and generalize the notion of a variable, as it depends on the protocol and hence may not always be private).

If there's an alternative way for you to do what you're doing now without `define_private_tensor` then I'm up for removing it.

Reply:
Feel like the best way would be to use `tfe.local_computation` but I haven't looked into how that would work yet.